src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -581,6 +618,142 @@ fn max_slice_length<'p, 'a: 'p, 'tcx: 'a, I>(
+    pub range: RangeInclusive<u128>,
+    pub ty: Ty<'tcx>,
+}
+
+impl<'tcx> Interval<'tcx> {
+    fn from_ctor(tcx: TyCtxt<'_, 'tcx, 'tcx>,
+                 ctor: &Constructor<'tcx>)
+                 -> Option<Interval<'tcx>> {
+        match ctor {
+            ConstantRange(lo, hi, end) => {
+                assert_eq!(lo.ty, hi.ty);
+                let ty = lo.ty;
+                if let Some(lo) = lo.assert_bits(ty) {
+                    if let Some(hi) = hi.assert_bits(ty) {
+                        // Perform a shift if the underlying types are signed,
+                        // which makes the interval arithmetic simpler.
+                        let (lo, hi) = Self::offset_sign(tcx, ty, lo..=hi, true);
+                        // Make sure the interval is well-formed.
+                        return if lo > hi || lo == hi && *end == RangeEnd::Excluded {
+                            None
+                        } else {
+                            let offset = (*end == RangeEnd::Excluded) as u128;
+                            Some(Interval { range: lo..=(hi - offset), ty })
+                        };
+                    }
+                }
+                None
+            }
+            ConstantValue(val) => {
+                let ty = val.ty;
+                if let Some(val) = val.assert_bits(ty) {
+                    let (lo, hi) = Self::offset_sign(tcx, ty, val..=val, true);
+                    Some(Interval { range: lo..=hi, ty })
+                } else {
+                    None
+                }
+            }
+            Single | Variant(_) | Slice(_) => {
+                None
+            }
+        }
+    }
+
+    fn offset_sign(tcx: TyCtxt<'_, 'tcx, 'tcx>,
+                   ty: Ty<'tcx>,
+                   range: RangeInclusive<u128>,
+                   encode: bool)
+                   -> (u128, u128) {
+        // We ensure that all integer values are contiguous: that is, that their
+        // minimum value is represented by 0, so that comparisons and increments/
+        // decrements on interval endpoints work consistently whether the endpoints
+        // are signed or unsigned.
+        let (lo, hi) = range.into_inner();
+        match ty.sty {
+            ty::TyInt(_) => {
+                // FIXME(49937): refactor these bit manipulations into interpret.
+                let bits = tcx.layout_of(ty::ParamEnv::reveal_all().and(ty))
+                                 .unwrap().size.bits() as u128;
+                let min = 1u128 << (bits - 1);
+                let shift = 1u128.overflowing_shl(bits as u32);
+                let mask = shift.0.wrapping_sub(1 + (shift.1 as u128));
+                if encode {
+                    let offset = |x: u128| x.wrapping_sub(min) & mask;
+                    (offset(lo), offset(hi))
+                } else {
+                    let offset = |x: u128| {
+                        interpret::sign_extend(tcx, x.wrapping_add(min) & mask, ty)
```

Comment:
You don't want a `sign_extend`, bit values are kept truncated in miri.

Reply:
I've added a `FIXME`. This shouldn't be necessary once `print_miri_value` does the sign extension, but it currently doesn't and it's much easier to do it here where we have a `TyCtxt`.

Reply:
FWIW this produces "wrong" values, that could result in weird behavior in miri.

Reply:
Okay, I'm fixing it.